### PR TITLE
Add domain allowlist to block SSRF via first-party proxy redirects

### LINKED ISSUE #414: SSRF via first-party proxy -- no target domain allowlist
## Summary

The `/first-party/proxy` endpoint proxies to arbitrary URLs (protected only by `tstoken` signature). `proxy_with_redirects` follows up to 4 redirects with no domain or IP range restriction, allowing SSRF to internal services if a signed URL redirects.

## Refs

- `crates/common/src/proxy.rs` lines 600-621 — `handle_first_party_proxy`
- `crates/common/src/proxy.rs` lines 463-582 — `proxy_with_redirects` follows redirects

## Recommendation

Validate redirect targets against an allowlist or block private IP ranges.

## Context

Production readiness audit — see #396

### PR DESCRIPTION
**Pull request by prk-Jr**
## Summary

- The first-party proxy followed redirects with no restriction on destination hosts. A signed URL pointing to an attacker-controlled origin could redirect to an internal service, enabling SSRF.
- Adds `proxy.allowed_domains` — an opt-in list of permitted redirect-target hostnames supporting exact match (`"example.com"`) and subdomain wildcard (`"*.example.com"`) with dot-boundary enforcement to prevent `evil-example.com` bypass.
- When the list is empty (the default), all redirect destinations are permitted — preserving backward compatibility for existing deployments.

## Changes

| File | Change |
| ---- | ------ |
| `crates/common/src/proxy.rs` | Add `is_host_allowed` helper; enforce allowlist on each redirect hop in `proxy_with_redirects`; 13 unit tests |
| `crates/common/src/settings.rs` | Add `allowed_domains: Vec<String>` field to `Proxy` struct with serde default |
| `trusted-server.toml` | Uncomment `[proxy]` section; add `allowed_domains` key with commented vendor examples |
| `.env.dev` | Document `TRUSTED_SERVER__PROXY__ALLOWED_DOMAINS` env var override |

## Closes

Closes #414

## Test plan

- [x] `cargo test --workspace`
- [x] `cargo clippy --all-targets --all-features -- -D warnings`
- [x] `cargo fmt --all -- --check`
- [ ] JS tests: `cd crates/js/lib && npx vitest run`
- [ ] JS format: `cd crates/js/lib && npm run format`
- [ ] WASM build: `cargo build --bin trusted-server-fastly --release --target wasm32-wasip1`
- [x] Manual testing via `fastly compute serve`

> Note: JS tests fail with `ERR_REQUIRE_ESM` due to a pre-existing `jsdom@28` / `html-encoding-sniffer` incompatibility on `main` — unrelated to this PR.

## Checklist

- [x] Changes follow [CLAUDE.md](/CLAUDE.md) conventions
- [x] No `unwrap()` in production code — use `expect("should ...")`
- [x] Uses `log` macros (not `println!`)
- [x] New code has tests
- [x] No secrets or credentials committed

### CODE REVIEW

File: docs/guide/first-party-proxy.md [hunk 1]
```diff
@@ -425,6 +425,47 @@ proxy_secret = "your-secure-random-secret"
...
+```toml
+[proxy]
+allowed_domains = [
+  "tracker.com",           # Exact match
+  "*.adserver.com",        # Wildcard: adserver.com and all subdomains
+  "*.trusted-cdn.net",
+]
+```
+
+**Semantics**: When a proxied request receives an HTTP redirect (301/302/303/307/308), the redirect target host is checked against `allowed_domains`. If the host does not match any pattern the redirect is blocked and a 403 error is returned.
```
Comment (aram356) [inline docs/guide/first-party-proxy.md]: 🔧 **wrench** — Status code mismatch: the code returns 502 (Bad Gateway) via `TrustedServerError::Proxy`, not 403 as stated here.

**Suggested fix:**
```
...the redirect is blocked and a 502 error is returned.
```

Alternatively, if 403 is the desired behavior, add a dedicated error variant that returns 403 for allowlist violations.

File: crates/trusted-server-core/src/settings.rs [hunk 2]
```diff
@@ -323,10 +334,28 @@ impl Default for Proxy {
...
 }
 
+impl Proxy {
+    /// Normalizes [`allowed_domains`] in place.
+    ///
+    /// Each entry is trimmed of surrounding whitespace and lowercased.
+    /// Empty entries (including those that were only whitespace) are removed.
+    /// This prevents blank patterns from making the list non-empty while
+    /// still never matching any host, which would silently block all redirects.
+    fn normalize(&mut self) {
```
Comment (aram356) [inline crates/trusted-server-core/src/settings.rs]: 🤔 **thinking** — Consider adding a `log::warn!` when `allowed_domains` is empty after normalization. Operators may not realize their deployment is in open mode (all redirect destinations permitted). A startup warning would improve operational visibility.

```rust
fn normalize(&mut self) {
    self.allowed_domains = self
        .allowed_domains
        .iter()
        .map(|s| s.trim().to_ascii_lowercase())
        .filter(|s| !s.is_empty())
        .collect();
    if self.allowed_domains.is_empty() {
        log::warn!("proxy.allowed_domains is empty: all redirect destinations are permitted");
    }
}
```

File: crates/trusted-server-core/src/proxy.rs [hunk 3]
```diff
@@ -479,6 +479,35 @@ fn append_synthetic_id(req: &Request, target_url_parsed: &mut url::Url) {
...
+}
+
+/// Returns `true` if `host` is permitted by `pattern`.
+///
+/// - `"example.com"` matches exactly `example.com`.
+/// - `"*.example.com"` matches `example.com` and any subdomain at any depth.
+///
+/// Comparison is case-insensitive. The wildcard check requires a dot boundary,
+/// so `"*.example.com"` does **not** match `"evil-example.com"`.
+fn is_host_allowed(host: &str, pattern: &str) -> bool {
```
Comment (aram356) [inline crates/trusted-server-core/src/proxy.rs]: 👍 **praise** — The `strip_suffix` + `ends_with('.')` approach for dot-boundary enforcement is clean and correct. It properly prevents `evil-example.com` from matching `*.example.com`, which is the most common allowlist bypass vector. Well-tested too.

File: crates/trusted-server-core/src/proxy.rs [hunk 4]
```diff
@@ -479,6 +479,35 @@ fn append_synthetic_id(req: &Request, target_url_parsed: &mut url::Url) {
     }
 }
 
+/// Returns `true` when a redirect to `host` should be followed.
+///
+/// When `allowed_domains` is empty every host is permitted (open mode).
+/// When non-empty the host must match at least one pattern via [`is_host_allowed`].
+fn redirect_is_permitted(allowed_domains: &[String], host: &str) -> bool {
```
Comment (aram356) [inline crates/trusted-server-core/src/proxy.rs]: 🌱 **seedling** — Per the project's newtype convention (CLAUDE.md), `allowed_domains: Vec<String>` could become an `AllowedDomains` newtype that encapsulates normalization and matching logic (`is_host_allowed`, `redirect_is_permitted`). Not needed for this PR, but would make the API more cohesive as the feature evolves.

File: crates/trusted-server-core/src/proxy.rs [hunk 5]
```diff
@@ -1834,4 +1875,177 @@ mod tests {
             body
         );
     }
+
+    // --- is_host_allowed ---
```
Comment (aram356) [inline crates/trusted-server-core/src/proxy.rs]: 🌱 **seedling** — The 18 allowlist tests are thorough for the documented cases. A few edge cases that could be added in a follow-up:

- **IP literal redirect target** (e.g., `is_host_allowed("169.254.169.254", "*.example.com")`) — the logic correctly blocks it, but an explicit test documents the guarantee
- **Empty Location header** in a redirect response

These are low risk since the current logic handles them correctly, but explicit tests would prevent regressions.

File: crates/trusted-server-core/src/proxy.rs [hunk 6]
```diff
@@ -479,6 +479,35 @@ fn append_synthetic_id(req: &Request, target_url_parsed: &mut url::Url) {
...
+    allowed_domains.is_empty() || allowed_domains.iter().any(|p| is_host_allowed(host, p))
+}
+
+/// Returns `true` if `host` is permitted by `pattern`.
+///
+/// - `"example.com"` matches exactly `example.com`.
+/// - `"*.example.com"` matches `example.com` and any subdomain at any depth.
+///
+/// Comparison is case-insensitive. The wildcard check requires a dot boundary,
+/// so `"*.example.com"` does **not** match `"evil-example.com"`.
```
[the remark this answers is not in the document]
Comment (prk-Jr) [reply crates/trusted-server-core/src/proxy.rs]: Fixed in `normalize()`: bare `"*"` entries are now removed with a `log::warn!` rather than silently falling through to always-false exact match. Added two settings tests: one for mixed list (`["*", "tracker.com"]` → `["tracker.com"]`) and one for solo `["*"]` → empty (open mode).

File: docs/guide/first-party-proxy.md [hunk 7]
```diff
@@ -425,6 +425,47 @@ proxy_secret = "your-secure-random-secret"
...
+
+**Wildcard matching**:
+
+| Pattern         | Matches                                             | Does not match     |
+| --------------- | --------------------------------------------------- | ------------------ |
+| `tracker.com`   | `tracker.com`                                       | `sub.tracker.com`  |
+| `*.tracker.com` | `tracker.com`, `sub.tracker.com`, `a.b.tracker.com` | `evil-tracker.com` |
+
+- The `*` prefix matches the base domain and any subdomain at any depth.
+- Matching is case-insensitive; entries are normalized to lowercase on startup.
```
[the remark this answers is not in the document]
Comment (prk-Jr) [reply docs/guide/first-party-proxy.md]: Updated. Wording is now: "This default exists solely for development convenience and **must be overridden in production**."

File: crates/trusted-server-core/src/proxy.rs [hunk 8]
```diff
@@ -584,6 +623,17 @@ async fn proxy_with_redirects(
...
+                next_host
+            );
+            return Err(Report::new(TrustedServerError::AllowlistViolation {
+                host: next_host.to_string(),
+            }));
+        }
+
         log::info!(
             "following redirect {} => {} (status {})",
             current_url,
```
[the remark this answers is not in the document]
Comment (prk-Jr) [reply crates/trusted-server-core/src/proxy.rs]: Fixed. A missing-host check was added before the allowlist check for redirect hops, returning `TrustedServerError::Proxy { message: "missing host in redirect location" }` instead of falling through to `AllowlistViolation { host: "" }`.

**Review by aram356 — APPROVED** [review]
## Summary

Solid SSRF mitigation for the first-party proxy redirect chain. The core matching logic (`is_host_allowed`) is correct, well-tested (13+ unit tests covering boundary bypass, IP literals, case sensitivity, empty lists), and cleanly scoped to only the first-party proxy path — integration proxies correctly default to open mode since they target hardcoded URLs.

CI is all green. No blocking issues found.

## Non-blocking

### 🤔 thinking

- **Wildcard semantics differ from `Rewrite::is_excluded`** — `*.example.com` matches the apex domain in the new `is_host_allowed` but not in `Rewrite::is_excluded`. Both behaviors are defensible for their use cases, but the inconsistency could confuse operators configuring both. Worth a note in docs if it comes up.

### 🌱 seedling

- **Consider deny-by-default in a future major version** — empty list = open mode is correct for backward compat, but a future breaking release could flip to require an explicit allowlist for production safety.

### 📌 out of scope

- **`handle_first_party_click` doesn't check allowlist** — correct for SSRF (no server-side request), but the click endpoint could still redirect a browser to an attacker-controlled URL via a signed URL. This is an open-redirect concern, not SSRF, and outside this PR's scope.

## CI Status

- cargo fmt: PASS
- cargo clippy: PASS
- cargo test: PASS
- vitest: PASS
- integration tests: PASS
- browser integration tests: PASS
- CodeQL: PASS
